File: impls/rust/src/step6_file.rs
use std::{cell::RefCell, collections::HashMap, io::Write, rc::Rc};

use rust2::{read_str, Env, MalError, MalFn, MalResult, MalVal, NS};

fn read(input: &str) -> Result<MalVal, MalError> {
    read_str(input)
}

fn eval_ast(ast: Rc<MalVal>, env: Rc<RefCell<Env>>) -> MalResult {
    match ast.as_ref() {
        MalVal::Symbol(symbol) => env
            .as_ref()
            .borrow()
            .get(symbol)
            .ok_or_else(|| MalError::Other(format!("'{symbol}' not found."))),
        MalVal::List(list) => {
            let mut buffer = Vec::new();
            for v in list {
                buffer.push(eval(v.clone(), env.clone())?);
            }
            Ok(Rc::new(MalVal::List(buffer)))
        }
        MalVal::Vector(vector) => {
            let mut buffer = Vec::new();
            for v in vector {
                buffer.push(eval(v.clone(), env.clone())?);
            }
            Ok(Rc::new(MalVal::Vector(buffer)))
        }
        MalVal::HashMap(hashmap) => {
            let mut buffer = HashMap::new();
            for (k, v) in hashmap {
                buffer.insert(k.clone(), eval(v.clone(), env.clone())?);
            }
            Ok(Rc::new(MalVal::HashMap(buffer)))
        }
        _ => Ok(ast),
    }
}

fn eval(ast: Rc<MalVal>, env: Rc<RefCell<Env>>) -> MalResult {
    let mut ast = ast;
    let mut env = env;
    loop {
        match ast.as_ref() {
            MalVal::List(list) if list.is_empty() => return Ok(ast),
            MalVal::List(list) => {
                if let MalVal::Symbol(symbol) = list[0].as_ref() {
                    match symbol.as_str() {
                        "def!" => match list[1].as_ref() {
                            MalVal::Symbol(symbol) => {
                                let v = eval(list[2].clone(), env.clone())?;
                                env.borrow_mut().set(symbol.clone(), v.clone());
                                return Ok(v);
                            }
                            _ => unreachable!(),
                        },
                        "let*" => {
                            let n_env = Rc::new(RefCell::new(Env::new(env)));
                            match list[1].as_ref() {
                                MalVal::List(binds) | MalVal::Vector(binds) => {
                                    let mut iter = binds.iter();
                                    while let Some(v) = iter.next() {
                                        match v.as_ref() {
                                            MalVal::Symbol(symbol) => {
                                                let value = eval(
                                                    iter.next().unwrap().clone(),
                                                    n_env.clone(),
                                                )?;
                                                n_env.borrow_mut().set(symbol.clone(), value);
                                            }
                                            _ => unreachable!(),
                                        }
                                    }
                                }
                                _ => unreachable!(),
                            }
                            env = n_env;
                            ast = list[2].clone();
                            continue;
                        }
                        "do" => {
                            let mut value = eval(list[1].clone(), env.clone())?;
                            for v in &list[2..] {
                                value = eval(v.clone(), env.clone())?;
                            }
                            return Ok(value);
                        }
                        "if" => {
                            let condition = eval(list[1].clone(), env.clone())?;
                            match condition.as_ref() {
                                MalVal::Nil | MalVal::Bool(false) => {
                                    if list.len() > 3 {
                                        ast = list[3].clone();
                                    } else {
                                        return Ok(Rc::new(MalVal::Nil));
                                    }
                                }
                                _ => ast = list[2].clone(),
                            };
                            continue;
                        }
                        "fn*" => {
                            let binds = match list[1].as_ref() {
                                MalVal::List(list) | MalVal::Vector(list) => list
                                    .iter()
                                    .map(|v| match v.as_ref() {
                                        MalVal::Symbol(symbol) => symbol.to_string(),
                                        _ => unreachable!(),
                                    })
                                    .collect::<Vec<_>>(),
                                _ => unreachable!(),
                            };
                            let body = list[2].clone();
                            return Ok(Rc::new(MalVal::Fn(Rc::new(MalFn::custom_func(
                                body, binds, env, eval,
                            )))));
                        }
                        _ => (),
                    }
                }
                let n_ast = eval_ast(ast, env)?;
                match n_ast.as_ref() {
                    MalVal::List(list) => match list[0].as_ref() {
                        MalVal::Fn(func) => match func.as_ref() {
                            MalFn::RegularFn(func) => return (func)(&list[1..]),
                            MalFn::MalFunc(func) => {
                                let mut n_env = Env::new(func.env.clone());
                                n_env.bind_expr(func.params.clone(), list[1..].to_vec());
                                ast = func.ast.clone();
                                env = Rc::new(RefCell::new(n_env));
                            }
                        },
                        _ => unreachable!(),
                    },
                    _ => unreachable!(),
                }
            }
            _ => return eval_ast(ast, env),
        }
    }
}

fn print(val: &MalVal) -> String {
    val.pr_str(true)
}

fn rep(input: &str, env: &Rc<RefCell<Env>>) -> Option<String> {
    match read(input).and_then(|ast| eval(Rc::new(ast), env.clone())) {
        Ok(v) => Some(print(v.as_ref())),
        Err(MalError::Continue) => None,
        Err(e) => Some(e.to_string()),
    }
}

fn main() {
    let mut env = Env::default();
    for (k, v) in NS {
        env.set(
            k.to_string(),
            Rc::new(MalVal::Fn(Rc::new(MalFn::RegularFn(Rc::new(v))))),
        );
    }
    let env = Rc::new(RefCell::new(env));
    let env_tmp = env.clone();
    env.as_ref().borrow_mut().set(
        "eval".to_string(),
        Rc::new(MalVal::Fn(Rc::new(MalFn::RegularFn(Rc::new(
            move |args| eval(args[0].clone(), env_tmp.clone()),
        ))))),
    );
    rep("(def! not (fn* (a) (if a false true)))", &env);
    rep(
        r#"(def! load-file (fn* (f) (eval (read-string (str "(do " (slurp f) "\nnil)")))))"#,
        &env,
    );

    let args = std::env::args().collect::<Vec<_>>();
    if args.len() > 1 {
        let mut iter = args.into_iter();
        iter.next();
        let filename = iter.next().unwrap();
        let init = Rc::new(MalVal::List(
            iter.map(|s| Rc::new(MalVal::String(s))).collect::<Vec<_>>(),
        ));
        env.as_ref().borrow_mut().set("*ARGV*".to_string(), init);
        let input = format!("(load-file \"{filename}\")");
        rep(&input, &env);
        return;
    }
    env.as_ref()
        .borrow_mut()
        .set("*ARGV*".to_string(), Rc::new(MalVal::List(Vec::new())));

    let mut buffer = String::new();
    loop {
        print!("user> ");
        std::io::stdout().flush().unwrap();
        let input = match std::io::stdin().read_line(&mut buffer) {
            Ok(0) => {
                println!();
                break;
            }
            Err(_) => break,
            Ok(_) => buffer.trim(),
        };
        if let Some(s) = rep(input, &env) {
            println!("{s}")
        }
        buffer.clear();
    }
}

File: impls/cpp/src/step6_file.cpp
#include <fmt/core.h>

#include <functional>
#include <string>
#include <string_view>

#include "core.h"
#include "env.h"
#include "linenoise.h"
#include "printer.h"
#include "reader.h"

std::shared_ptr<MalType> Eval(std::shared_ptr<MalType> ast,
                              std::shared_ptr<Env> env);
std::shared_ptr<MalType> EvalAst(const std::shared_ptr<MalType>& ast,
                                 std::shared_ptr<Env> env) {
    if (auto s = std::dynamic_pointer_cast<Symbol>(ast)) {
        return env->Get(**s);
    } else if (auto v = std::dynamic_pointer_cast<Vector>(ast)) {
        auto vector = std::make_shared<Vector>();
        for (auto& item : **v) (*vector)->emplace_back(Eval(item, env));
        return vector;
    } else if (auto l = std::dynamic_pointer_cast<List>(ast)) {
        auto list = std::make_shared<List>();
        for (auto& v : **l) (*list)->emplace_back(Eval(v, env));
        return list;
    } else if (auto h = std::dynamic_pointer_cast<HashMap>(ast)) {
        auto map = std::make_shared<HashMap>();
        for (auto& [k, v] : **h) (*map)->emplace(k, Eval(v, env));
        return map;
    }
    return ast;
}

auto Read(std::string_view str) { return ReadStr(str); }

std::shared_ptr<MalType> Eval(std::shared_ptr<MalType> ast,
                              std::shared_ptr<Env> env) {
    while (true) {
        if (auto l = std::dynamic_pointer_cast<List>(ast)) {
            if ((*l)->empty()) return ast;
            auto it = (*l)->begin();
            if (auto s = std::dynamic_pointer_cast<Symbol>(*it)) {
                if (const auto& action = **s; action == "def!") {
                    auto symbol = **std::dynamic_pointer_cast<Symbol>(*(++it));
                    auto value = Eval(*(++it), env);
                    env->Set(std::move(symbol), value);
                    return value;
                } else if (action == "let*") {
                    env = std::make_shared<Env>(env);
                    auto bindings =
                        std::dynamic_pointer_cast<Sequence>(*(++it));
                    for (auto it = (*bindings)->begin();
                         it + 1 < (*bindings)->end(); it += 2) {
                        auto symbol = **std::dynamic_pointer_cast<Symbol>(*it);
                        auto value = Eval(*(it + 1), env);
                        env->Set(std::move(symbol), std::move(value));
                    }
                    ast = *(++it);
                    continue;
                } else if (action == "do") {
                    std::shared_ptr<MalType> res;
                    for (++it; it != (*l)->end(); it++) res = Eval(*it, env);
                    return res;
                } else if (action == "if") {
                    auto condition = Eval(*(++it), env);
                    if (std::dynamic_pointer_cast<False>(condition) ||
                        std::dynamic_pointer_cast<Nil>(condition)) {
                        if ((*l)->size() < 4) return MalType::Nil;
                        ast = *(it + 2);
                        continue;
                    } else {
                        ast = *(++it);
                        continue;
                    }
                } else if (action == "fn*") {
                    auto binds = std::dynamic_pointer_cast<Sequence>(*(++it));
                    auto ast = *(++it);
                    std::vector<std::string> params((*binds)->size());
                    std::transform(
                        (*binds)->begin(), (*binds)->end(), params.begin(),
                        [](std::shared_ptr<MalType> arg) -> std::string {
                            return **std::dynamic_pointer_cast<Symbol>(arg);
                        });
                    UserFunc::Callback f =
                        [](std::span<std::shared_ptr<MalType>> args,
                           const UserFunc& func) {
                            auto new_env = std::make_shared<Env>(
                                func.get_params(), args, func.get_env());
                            return Eval(func.get_ast(), new_env);
                        };
                    return std::make_shared<UserFunc>(ast, params, env, f);
                }
            }
            auto list = std::dynamic_pointer_cast<List>(EvalAst(l, env));
            it = (*list)->begin();
            auto args = std::span{it + 1, (*list)->end()};
            if (auto func = std::dynamic_pointer_cast<UserFunc>(*(it))) {
                ast = func->get_ast();
                auto& params = func->get_params();
                env = std::make_shared<Env>(
                    std::span{params.begin(), params.end()}, args,
                    func->get_env());
                continue;
            }
            return (*std::dynamic_pointer_cast<MalFunc>(*(it)))(args);
        }
        return EvalAst(ast, env);
    }
}

std::string Print(auto ast) { return PrStr(ast, true); }
std::string Rep(std::string_view str, std::shared_ptr<Env> env) {
    try {
        return Print(Eval(Read(str), env));
    } catch (std::string_view err) {
        fmt::print("{}", err);
        return "";
    } catch (std::string err) {
        fmt::print("{}", err);
        return "";
    }
}

int main(int argc, const char* argv[]) {
    linenoiseHistorySetMaxLen(30);
    linenoiseHistoryLoad("history.txt");

    auto env = std::make_shared<Env>();

    for (auto& [k, v] : getNS()) {
        auto f = std::make_shared<BaseFunc>(v);
        env->Set(std::string{k}, f);
    }

    env->Set("eval", std::make_shared<BaseFunc>(
                         [env = std::weak_ptr<Env>(env)](
                             std::span<std::shared_ptr<MalType>> args) {
                             return Eval(args[0], env.lock());
                         }));

    Rep("(def! not (fn* (a) (if a false true)))", env);
    Rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) "
        "\"\nnil)\")))))",
        env);

    auto list = std::make_shared<List>();
    if (argc > 1) {
        for (int i = 2; i < argc; i++)
            (*list)->push_back(std::make_shared<String>(argv[i]));
        env->Set("*ARGV*", std::move(list));
        Rep(fmt::format("(load-file \"{}\")", argv[1]), env);
        return 0;
    }
    env->Set("*ARGV*", std::move(list));

    char* line;
    while ((line = linenoise("user> ")) != NULL) {
        if (line[0] == '\0') continue;
        linenoiseHistoryAdd(line);
        try {
            fmt::print("{}\n", Rep(line, env));
        } catch (std::nullptr_t) {
            linenoiseFree(line);
            continue;
        }
        linenoiseFree(line);
    }
    linenoiseHistorySave("history.txt");
    return 0;
}
